cmd/warm-backend-s3.go

```diff
@@ -0,0 +1,127 @@
+/*
+ * MinIO Cloud Storage, (C) 2021 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"time"
+
+	minio "github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/credentials"
+	"github.com/minio/minio/pkg/madmin"
+)
+
+type warmBackendS3 struct {
+	client       *minio.Client
+	core         *minio.Core
+	Bucket       string
+	Prefix       string
+	StorageClass string
+}
+
+func (s3 *warmBackendS3) ToObjectError(err error, params ...string) error {
+	object := ""
+	if len(params) >= 1 {
+		object = params[0]
+	}
+
+	return ErrorRespToObjectError(err, s3.Bucket, s3.getDest(object))
+}
+
+func (s3 *warmBackendS3) getDest(object string) string {
+	destObj := object
+	if s3.Prefix != "" {
+		destObj = fmt.Sprintf("%s/%s", s3.Prefix, object)
+	}
+	return destObj
+}
+
+func (s3 *warmBackendS3) Put(ctx context.Context, object string, r io.Reader, length int64) error {
+	_, err := s3.client.PutObject(ctx, s3.Bucket, s3.getDest(object), r, length, minio.PutObjectOptions{StorageClass: s3.StorageClass})
+	return s3.ToObjectError(err, object)
+}
+
+func (s3 *warmBackendS3) Get(ctx context.Context, object string, opts warmBackendGetOpts) (io.ReadCloser, error) {
+	gopts := minio.GetObjectOptions{}
+
+	if opts.startOffset >= 0 && opts.length >= 0 {
```

Comment:
offset and length has already been calculated by NewGetObjectReader, this is &&

Reply:
> offset and length has already been calculated by NewGetObjectReader, this is &&

Okay.

What  I suggest is we void this use case, startOffset = 0 and length = 0, because it will lead gopts.SetRange(0, -1)

Reply:
accounted for startOffset = 0 and length = 0 in the above condition